tests/kernel/mem_protect/userspace/src/main.c

```diff
@@ -488,6 +516,164 @@ static void read_kobject_user_pipe(void)
+	zassert_unreachable("Create higher priority user thread greater"
+			" than current thread did not fault\n");
+}
+
+struct k_mem_domain app0_domain;
+u8_t MEM_DOMAIN_ALIGNMENT app0_buf0[MEM_REGION_ALLOC];
+u8_t MEM_DOMAIN_ALIGNMENT app0_buf1[MEM_REGION_ALLOC];
+
+K_MEM_PARTITION_DEFINE(app0_part0, app0_buf0, sizeof(app0_buf0),
+		K_MEM_PARTITION_P_RW_U_RW);
+
+K_MEM_PARTITION_DEFINE(app0_part1, app0_buf1, sizeof(app0_buf1),
+		K_MEM_PARTITION_P_RW_U_RW);
+
+struct k_mem_partition *app0_parts[] = {
+	&app0_part0,
+	&app0_part1
+};
+
+struct k_mem_domain app1_domain;
+u8_t MEM_DOMAIN_ALIGNMENT app1_buf0[MEM_REGION_ALLOC];
+u8_t MEM_DOMAIN_ALIGNMENT app1_buf1[MEM_REGION_ALLOC];
+
+K_MEM_PARTITION_DEFINE(app1_part0, app1_buf0, sizeof(app1_buf0),
+		K_MEM_PARTITION_P_RW_U_RW);
+
+K_MEM_PARTITION_DEFINE(app1_part1, app1_buf1, sizeof(app1_buf1),
+		K_MEM_PARTITION_P_RW_U_RW);
+
+struct k_mem_partition *app1_parts[] = {
+	&app1_part0,
+	&app1_part1
+};
+
+__kernel struct k_thread utid, ktid;
+K_THREAD_STACK_DEFINE(ustack, MEM_DOMAIN_STACK_SIZE);
+K_THREAD_STACK_DEFINE(kstack, MEM_DOMAIN_STACK_SIZE);
+
+/***************************************************************************/
+static void kthread_entry(void *p1, void *p2, void *p3)
+{
+	/* Add current thread to memory domain app0_domain*/
+	k_mem_domain_add_thread(&app0_domain, k_current_get());
+
+	/* Add current thread to another memory domain app1_domain
+	 * This should cause a assertion failure, provided CONFIG_ASSERT=y
+	 * Cannot add a thread to two memory domains
+	 */
+	k_mem_domain_add_thread(&app1_domain, k_current_get());
+
+		/* These lines of code should not be reached */
+	zassert_unreachable("Adding a thread to multiple memory domains"
+			" and accessing them did not fault\n");
+}
+
+/* Verify that a thread can belong to a single memory domain
+ * at any point of time
+ * Try to add a thread to two memory domains at a time
+ */
+static void thread_in_multiple_domains(void)
+{
+	k_mem_domain_init(&app0_domain, ARRAY_SIZE(app0_parts), app0_parts);
+	k_mem_domain_init(&app1_domain, ARRAY_SIZE(app1_parts), app1_parts);
+
+	k_thread_create(&kthread_thread, kthread_stack, STACKSIZE,
+			(k_thread_entry_t)kthread_entry, NULL, NULL, NULL,
+			K_PRIO_PREEMPT(1), 0, K_NO_WAIT);
+
+	k_sem_take(&sync_sem, SEM_TIMEOUT);
+}
+
+/**************************************************************************/
+/* Test to validate that Threads in the same memory domain have
+ * the same access permissions to the memory partitions
+ * belonging to the memory domain
+ */
+struct k_mem_domain app_domain;
+u8_t MEM_DOMAIN_ALIGNMENT app_buf[MEM_REGION_ALLOC];
+K_MEM_PARTITION_DEFINE(app_part, app_buf, sizeof(app_buf),
+		K_MEM_PARTITION_P_RO_U_RO);
+struct k_mem_partition *app_parts[] = {
+	    &app_part
+};
+
+static void kentry(void)
+{
+	printk("Read the partition app_buf[0]=%d\n", app_buf[0]);
+
+	/* Write to the partition, this should generate fault */
+	app_buf[0] = 10;
+
```

Comment:
Same problem as in https://github.com/zephyrproject-rtos/zephyr/pull/5197#discussion_r166453992